### PR TITLE
Rob's comments and update variable view when done debugging

### PR DESCRIPTION
**Pull request by DavidKutu**
For #7323 and Rob's comments [here](https://github.com/microsoft/vscode-jupyter/pull/7502#discussion_r708452082)

-   [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [ ] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/client/debugger/jupyter/debugger.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    private rejectFunc?: (reason?: Error) => void;
+
+    readonly session: Promise<DebugSession>;
+
+    constructor(
+        public readonly document: NotebookDocument,
+        public readonly config: DebugConfiguration,
+        options?: DebugSessionOptions
+    ) {
+        this.session = new Promise<DebugSession>((resolve, reject) => {
```
Comment (rchiodo) [inline src/client/debugger/jupyter/debugger.ts:18]: You might try using a deferred object for this idea of remembering the promise resolve/reject. 

If you create a deferred object, you can implement resolve/reject with it.

**Review by rchiodo — APPROVED** [review]

**Review by IanMatthewHuff — APPROVED** [review]

File: src/client/datascience/jupyter/debuggerVariables.ts [hunk 2]
```diff
@@ -452,6 +415,65 @@ export class DebuggerVariables extends DebugLocationTracker
...
+                    threadId,
+                    startFrame: 0,
+                    levels: 1
+                });
+
+                //  Call scopes
+                const sf = stResponse.stackFrames[0];
+                const mode = this.debuggingManager.getDebugMode(doc);
+                // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                let scopesResponse: any | undefined;
```
Comment (roblourens) [inline src/client/datascience/jupyter/debuggerVariables.ts:438]: Why `any`?

File: src/client/datascience/jupyter/debuggerVariables.ts [hunk 3]
```diff
@@ -452,6 +415,65 @@ export class DebuggerVariables extends DebugLocationTracker
...
+                    const docURI = path.basename(doc.uri.toString());
+                    if (sf.source && sf.source.path && sf.source.path.includes(docURI)) {
+                        scopesResponse = await session.customRequest('scopes', { frameId: sf.id });
+                    }
+                }
+
+                // Call variables
+                if (scopesResponse) {
+                    // Keep track of variablesReference because "hover" requests also try to update variables
+                    const newVariablesReference = scopesResponse.scopes[0].variablesReference;
```
Comment (roblourens) [inline src/client/datascience/jupyter/debuggerVariables.ts:457]: Are you sure you only want to update the first scope?

File: src/client/datascience/jupyter/debuggerVariables.ts [hunk 4]
```diff
@@ -300,10 +267,6 @@ export class DebuggerVariables extends DebugLocationTracker
 
             this.updateVariables(undefined, message as DebugProtocol.VariablesResponse);
```
Comment (roblourens) [inline src/client/datascience/jupyter/debuggerVariables.ts:268]: Can't this stuff be deleted, it seems like you copied the code to line 469

File: src/client/debugger/types.ts [hunk 5]
```diff
@@ -16,39 +16,40 @@ export interface IKernelDebugAdapter {
...
     getDebugCell(notebook: NotebookDocument): NotebookCell | undefined;
 }
 
-export interface DebuggingDelegate {
+export interface IDebuggingDelegate {
     /**
      * Called for every event sent from the debug adapter to the client. Returns true to signal that sending the message is vetoed.
      */
-    willSendEvent(msg: DebugProtocolMessage): Promise<boolean>;
+    onWillSendEvent(msg: DebugProtocolMessage): Promise<boolean>;
```
Comment (roblourens) [inline src/client/debugger/types.ts:30]: Usually I feel that events start with `on...`, this is a method and not an event. I don't know if your codebase has the same pattern, it's up to you

Comment (DavidKutu) [reply src/client/debugger/types.ts]: true, that would be confusing, I'll put it back

File: src/client/datascience/jupyter/debuggerVariables.ts (hunk 4 above)
Comment (DavidKutu) [reply src/client/datascience/jupyter/debuggerVariables.ts:268]: I thought it was needed, but after testing, its not, I'll remove it.

File: src/client/datascience/jupyter/debuggerVariables.ts (hunk 3 above)
Comment (DavidKutu) [reply src/client/datascience/jupyter/debuggerVariables.ts:457]: no hahah

File: src/client/debugger/jupyter/debugControllers.ts [hunk 6]
```diff
@@ -41,7 +41,7 @@ export class DebugCellController implements DebuggingDelegate {
     }
 }
 
-export class RunByLineController implements DebuggingDelegate {
+export class RunByLineController implements IDebuggingDelegate {
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debugControllers.ts:44]: Curious, do we need all of these interfaces? Why can't we just use classes

File: src/client/debugger/jupyter/debugger.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    reject(reason: Error) {
+        if (this.rejectFunc) {
+            this.rejectFunc(reason);
+        }
+    }
+
+    async stop() {
+        void debug.stopDebugging(await this.session);
```
Comment (DonJayamanne) [inline src/client/debugger/jupyter/debugger.ts:39]: Why is this async if we're ignoring the promise returned by `debug.stopDebuggin`

File: src/client/debugger/jupyter/debugControllers.ts (hunk 6 above)
Comment (DavidKutu) [reply src/client/debugger/jupyter/debugControllers.ts:44]: We were following the extension patterns.

File: src/client/debugger/jupyter/debugger.ts (hunk 7 above)
Comment (DavidKutu) [reply src/client/debugger/jupyter/debugger.ts:39]: because we have to wait to get the session

**Pull request merged**